python_modules/dagster/dagster/_core/execution/backfill.py

```diff
@@ -508,3 +518,50 @@ def from_asset_graph_subset(
             title=title,
             description=description,
         )
+
+
+def cancel_backfill_runs_and_cancellation_complete(
+    instance: "DagsterInstance", backfill_id: str
+) -> Iterator[Union[None, bool]]:
+    """Cancels MAX_RUNS_CANCELED_PER_ITERATION runs associated with the backfill_id. Ensures that
+    all runs for the backfill are in a terminal state before indicating that the backfill can be marked
+    CANCELED.
+    Yields a boolean indicating the backfill can be considered canceled (ie all runs are canceled).
+    """
+    if not instance.run_coordinator:
+        check.failed("The instance must have a run coordinator in order to cancel runs")
+
+    # Query for cancelable runs, enforcing a limit on the number of runs to cancel in an iteration
+    # as canceling runs incurs cost
+    runs_to_cancel_in_iteration = instance.run_storage.get_run_ids(
+        filters=RunsFilter(
+            statuses=CANCELABLE_RUN_STATUSES,
+            tags={
+                BACKFILL_ID_TAG: backfill_id,
+            },
+        ),
+        limit=MAX_RUNS_CANCELED_PER_ITERATION,
+    )
+
+    yield None
+
+    if runs_to_cancel_in_iteration:
+        # since we are canceling some runs in this iteration, we know that there is more work to do.
+        # Either cancelling more runs, or waiting for the canceled runs to get to a terminal state
+        work_done = False
+        for run_id in runs_to_cancel_in_iteration:
+            instance.run_coordinator.cancel_run(run_id)
+            yield None
+    else:
+        # If there are no runs to cancel, check if there are any runs still in progress. If there are,
+        # then we want to wait for them to reach a terminal state before the backfill is marked CANCELED.
+        run_waiting_to_cancel = instance.get_run_ids(
+            RunsFilter(
+                tags={BACKFILL_ID_TAG: backfill_id},
+                statuses=IN_PROGRESS_RUN_STATUSES,
```

Comment:
@clairelin135  what do you think about changing this to `NOT_FINISHED_STATUSES` so that we detect any runs in STARTING or NOT_STARTED states and wait to mark the backfill canceled until those runs move to a status where we can cancel them?

Reply:
that makes sense - we probably should have done that in the first place